core/src/main/scala/org/apache/spark/executor/ProcfsBasedSystems.scala

```diff
@@ -0,0 +1,222 @@
+      // and it will have a exit code of 2 if there is no child process
+      if (exitCode != 0 && exitCode > 2) {
+        val cmd = builder.command().toArray.mkString(" ")
+        logWarning(s"Process $cmd" +
+          s" exited with code $exitCode, with stderr:" + s"${process.getErrorStream} ")
+        throw new SparkException(s"Process $cmd exited with code $exitCode")
+      }
+      childPidsInInt
+    } catch {
+      case e: Exception => logWarning("Exception when trying to compute process tree." +
+        " As a result reporting of ProcessTree metrics is stopped.", e)
+        isAvailable = false
+        return mutable.ArrayBuffer.empty[Int]
+    }
+  }
+
+  def computeProcessInfo(pid: Int): Unit = {
+    /*
+   * Hadoop ProcfsBasedProcessTree class used regex and pattern matching to retrive the memory
+   * info. I tried that but found it not correct during tests, so I used normal string analysis
+   * instead. The computation of RSS and Vmem are based on proc(5):
+   * http://man7.org/linux/man-pages/man5/proc.5.html
+   */
+    try {
+      val pidDir = new File(procfsDir, pid.toString)
+      Utils.tryWithResource( new InputStreamReader(
+        new FileInputStream(
+          new File(pidDir, procfsStatFile)), Charset.forName("UTF-8"))) { fReader =>
+        Utils.tryWithResource( new BufferedReader(fReader)) { in =>
+          val procInfo = in.readLine
+          val procInfoSplit = procInfo.split(" ")
+          if (procInfoSplit != null) {
+            val vmem = procInfoSplit(22).toLong
+            val rssPages = procInfoSplit(23).toLong
+            if (procInfoSplit(1).toLowerCase(Locale.US).contains("java")) {
+              allMetrics = allMetrics.copy(
+                jvmVmemTotal = allMetrics.jvmVmemTotal + vmem,
+                jvmRSSTotal = allMetrics.jvmRSSTotal + (rssPages*pageSize)
+              )
+            }
+            else if (procInfoSplit(1).toLowerCase(Locale.US).contains("python")) {
+              allMetrics = allMetrics.copy(
+                pythonVmemTotal = allMetrics.pythonVmemTotal + vmem,
+                pythonRSSTotal = allMetrics.pythonRSSTotal + (rssPages*pageSize)
+              )
+            }
+            else {
+              allMetrics = allMetrics.copy(
+                otherVmemTotal = allMetrics.otherVmemTotal + vmem,
+                otherRSSTotal = allMetrics.otherRSSTotal + (rssPages*pageSize)
+              )
+            }
+          }
+        }
+      }
+    } catch {
+      case f: FileNotFoundException => logWarning("There was a problem with reading" +
+        " the stat file of the process. ", f)
+    }
+  }
+
+  private[spark] def computeAllMetrics(): ProcfsBasedSystemsMetrics = {
+    if (!isAvailable) {
+      return ProcfsBasedSystemsMetrics(0, 0, 0, 0, 0, 0)
+    }
+    computeProcessTree
+    val pids = ptree.keySet
+    allMetrics = ProcfsBasedSystemsMetrics(0, 0, 0, 0, 0, 0)
+    for (p <- pids) {
+      computeProcessInfo(p)
```

Comment:
the state used here is a little trickier than it needs to be. 

 `computeProcessTree` is updating a member variable, even though its only used locally -- it would be easier to follow if instead it just returned the process tree, and then you passed it around.  Also I dont' think you actually care about the tree, just the set of pids?

similarly for `allMetrics`.  it doesn't really need to be a member variable, since its use is entirely contained within this function, you could just pass it around.

```scala
val pids = discoverPids()
val allMetrics = ...
for (p <- pids) {
  allMetrics = updateMetricsForProcess(allMetrics, p)
}
```

Reply:
The tree was there in case we want to do some other stuff with it, but I guess we can have a tree structure when we actually need it. Right now as you mentioned we don't need it. So I will change it.
the allMetrics was there for testing, but I can change the test anyway.